### PR TITLE
Extra array functions

### PR DESCRIPTION
**Pull request by chasent**
Thank you for this very helpful library!
I have added some functions I thought would have been useful when I first picked it up.

### CODE REVIEW

**Comment by gcanti** [comment]
Thanks @chasent, `flatMap` is already there (is named `chain` though)

**Comment by raveclassic** [comment]
@gcanti Awesome! Thanks!
Do you think it would be better to use native `find` and `fromNullable` for `find` (`findFirst`) and `for cycle`  for `findLast` for the sake of performance?

File: src/Array.ts [hunk 1]
```diff
@@ -42,6 +42,13 @@ export const concat = <A>(x: Array<A>) => (y: Array<A>): Array<A> => x.concat(y)
 
 export const map = <A, B>(f: (a: A) => B, fa: Array<A>): Array<B> => fa.map(f)
 
+/**
+ * Apply a function that returns an array to each element in an array,
+ * flattening the results into a new array.
+ */
+export const flatMap = <A, B>(f: (a: A) => Array<B>) => (fa: Array<A>): Array<B> =>
```
Comment (raveclassic) [inline src/Array.ts]: There is already a `chain` function

File: src/Array.ts [hunk 2]
```diff
@@ -182,6 +189,12 @@ export const findIndex = <A>(predicate: Predicate<A>) => (as: Array<A>): Option<
   return option.none
 }
 
+/** Find the first element which satisfies a predicate function */
+export const findFirst = <A>(predicate: Predicate<A>) => (as: Array<A>): Option<A> => head(filter(predicate)(as))
```
Comment (raveclassic) [inline src/Array.ts]: `findFirst` can be replaced with more efficient native `find` wrapped with `fromNullable`

File: src/Array.ts [hunk 3]
```diff
@@ -182,6 +189,12 @@ export const findIndex = <A>(predicate: Predicate<A>) => (as: Array<A>): Option<
...
+
+/** Find the last element which satisfies a predicate function */
+export const findLast = <A>(predicate: Predicate<A>) => (as: Array<A>): Option<A> => last(filter(predicate)(as))
```
Comment (raveclassic) [inline src/Array.ts]: `findLast` can be replaced with more efficient reversed `for`-cycle also wrapped with `fromNullable`

**Comment by chasent** [comment]
Welp, I feel stupid for not noticing `chain.flatten(array)`!
I've addressed the feedback given.

**Comment by raveclassic** [comment]
@chasent Great, thanks!

**Pull request merged**